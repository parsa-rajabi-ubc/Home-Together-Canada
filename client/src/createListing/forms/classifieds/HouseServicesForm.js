/**
 * @Author:     Alex Qin
 * @Created:    2021.02.12
 *
 * @Description: House Services Form
 *
 */

import React, {useEffect, useState} from 'react';
import PropTypes from "prop-types";
import TextArea from "../../../common/forms/TextArea";
import LargeTextArea from "../../../common/forms/LargeTextArea";
import SubmitButton from "../../../common/forms/SubmitButton";
import {SHORT_DESC_CHAR_COUNT} from "../../../common/constants/listingsConstants";
import {
    checkIfErrorsExistInMapping,
    validateInput
} from "../../../registration/registrationUtils";
import {HOUSE_YARD_TEXT as TEXT} from "./constants/ClassifiedsListingText";
import Tooltip from "../../../common/forms/Tooltip";
import {CREATE_LISTING_MEMBER_SHARE_HOME as ToolTipText} from "../../../common/constants/TooltipText";
import UploadImage from "../../../common/forms/UploadImage";

const HouseServicesForm = (props) => {
    const { onSubmit } = props;

    const [title, setTitle] = useState(undefined);
    const [shortDescription, setShortDescription] = useState(undefined);
    const [fullDescription, setFullDescription] = useState(undefined);
    const [rateAndFee, setRateAndFee] = useState(undefined);
    const [picture, setPicture] = useState(undefined);

    const [titleError, setTitleError] = useState(undefined);
    const [shortDescriptionError, setShortDescriptionError] = useState(undefined);
    const [fullDescriptionError, setFullDescriptionError] = useState(undefined);
    const [rateAndFeeError, setRateAndFeeError] = useState(undefined);

    useEffect(() => {
        title !== undefined && validateInput(title, setTitleError);
    }, [title]);
    useEffect(() => {
        shortDescription !== undefined && validateInput(shortDescription, setShortDescriptionError);
    }, [shortDescription]);
    useEffect(() => {
        fullDescription !== undefined && validateInput(fullDescription, setFullDescriptionError);
    }, [fullDescription]);
    useEffect(() => {
        rateAndFee !== undefined && validateInput(rateAndFee, setRateAndFee);
    }, [rateAndFee]);

    function handleImageUpload(e) {
        setPicture(e.target.files[0]);
    }

    const isFormValid = () => {

        const errors = {
            title: false,
            shortDes: false,
            fullDes: false,
            rateFee: false
        }

        errors.title = validateInput(title, setTitleError);
        errors.shortDes = validateInput(shortDescription, setShortDescriptionError);
        errors.fullDes = validateInput(fullDescription, setFullDescriptionError);
        errors.rateFee = validateInput(rateAndFee,setRateAndFeeError);

        return !(checkIfErrorsExistInMapping(errors));
    }

    //function for input checks on submit
    function onCreateListing() {

        if (!isFormValid()) {
            console.log("Form is invalid")
            // form is invalid
        }
        // else form is valid and proceed to make request
        else {
            onSubmit();
        }
    }

    return (
        <div>
            <div className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl">
                <div className="grid grid-cols-2 gap-6">
                    <div className="col-span-3 sm:col-span-2">

                        <h1 className={"page-title mb-5"}> {TEXT.form_title} </h1>

                        <TextArea
                            className={`${titleError && "border-red-500"} input`}
                            label={TEXT.title}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setTitle(e.target.value)}
                        />

                        <div className={"grid grid-cols-9 gap-x-6"}>

                            <section className={"col-start-1 col-end-5"}>

                                <TextArea
                                    className={`${shortDescriptionError && "border-red-500"} input`}
                                    label={TEXT.short_des + " (" + SHORT_DESC_CHAR_COUNT + " Characters)"}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setShortDescription(e.target.value)}
                                    charLimit={SHORT_DESC_CHAR_COUNT}
                                />
                            </section>

                            <section className={"col-start-1 col-end-5"}>
                                <TextArea
                                    className={`${rateAndFeeError && "border-red-500"} input`}
                                    label={TEXT.rateAndFees}
                                    labelClassName={"label"}
                                    required={true}
                                    onChange={(e) => setRateAndFee(e.target.value)}
                                />
                            </section>


                        </div>

                        <LargeTextArea
                            className={`${fullDescriptionError && "border-red-500"} input`}
                            rows={"6"}
                            label={TEXT.full_des}
                            labelClassName={"label"}
                            required={true}
                            onChange={(e) => setFullDescription(e.target.value)}
                        />

                        <label className="label"> Pictures </label>
                        <Tooltip
                            text={ToolTipText.PHOTOS}
                            toolTipID={"UploadPhotos"}
                        />
                        <UploadImage handleImageUpload={handleImageUpload}/>

                    </div>
                </div>
            </div>
            <SubmitButton className={"btn btn-green form-btn w-1/2"} onClick={onCreateListing}
                          onSubmit={onCreateListing}/>
        </div>
    )

}
HouseServicesForm.propTypes = {
    onSubmit: PropTypes.func.isRequired
}

export default HouseServicesForm;