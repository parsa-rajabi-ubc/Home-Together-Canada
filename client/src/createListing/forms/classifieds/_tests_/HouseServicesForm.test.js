/**
 * @Author:     Alex Qin
 * @Created:    2021.2.12
 *
 * @Description: House Services Form Test
 *
 */

import React from 'react';
import renderer from  'react-test-renderer'
import HouseServicesForm from "../HouseServicesForm";

jest.mock("uuid", () => ({
            v4: () => "00000000-0000-0000-0000-000000000000"}
    )
);

describe('HouseServicesForm', () => {
    describe('Snapshot test', () => {
        it("should render correctly regardless of properties", () => {
            // given
            const props = {
                onSubmit: jest.fn()
            };
            //when
            const component = renderer.create(<HouseServicesForm {...props}/>).toJSON();
            //then
            expect(component).toMatchSnapshot();
        });
    })
})