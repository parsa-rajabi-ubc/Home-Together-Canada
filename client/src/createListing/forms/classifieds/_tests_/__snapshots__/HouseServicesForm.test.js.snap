// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HouseServicesForm Snapshot test should render correctly regardless of properties 1`] = `
<div>
  <div
    className="col-start-1 col-end-7 py-5 px-5 m-6 bg-white shadow-lg rounded-xl"
  >
    <div
      className="grid grid-cols-2 gap-6"
    >
      <div
        className="col-span-3 sm:col-span-2"
      >
        <h1
          className="page-title mb-5"
        >
           
          House & Yard Services
           
        </h1>
        <section>
          <label
            className="label"
          >
            Descriptive Title
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <input
            className="undefined input"
            disabled={false}
            onChange={[Function]}
            type="text"
          />
        </section>
        <div
          className="grid grid-cols-9"
        >
          <section
            className="col-start-1 col-end-5"
          >
            <section>
              <label
                className="label"
              >
                Short Description (50 Characters)
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="undefined input"
                disabled={false}
                maxLength="50"
                onChange={[Function]}
                type="text"
              />
            </section>
          </section>
          <section
            className="col-start-6 col-end-10"
          >
            <section>
              <label
                className="label"
              >
                Rates and Fees
              </label>
              <svg
                className="inline align-top text-red-600"
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
                  fill="currentColor"
                />
              </svg>
              <input
                className="undefined input"
                disabled={false}
                onChange={[Function]}
                type="text"
              />
            </section>
          </section>
        </div>
        <div>
          <label
            className="label"
          >
            Full Description
          </label>
          <svg
            className="inline align-top text-red-600"
            fill="none"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 6H13V10.079L16.3413 7.73938L17.4885 9.37768L13.7434 12L17.4885 14.6223L16.3413 16.2606L13 13.921V18H11V13.921L7.65864 16.2606L6.51148 14.6223L10.2565 12L6.51147 9.37769L7.65863 7.73938L11 10.079V6Z"
              fill="currentColor"
            />
          </svg>
          <textarea
            className="undefined undefined input"
            cols="50"
            onChange={[Function]}
            rows="6"
          />
        </div>
        <label
          className="label"
        >
           Pictures 
        </label>
        <span>
          <span
            data-for="UploadPhotos"
            data-tip={true}
          >
            <svg
              className="inline ml-1 mb-1 text-blue-500 hover:text-blue-400"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M2 0a2 2 0 00-2 2v12a2 2 0 002 2h12a2 2 0 002-2V2a2 2 0 00-2-2H2zm4.57 6.033H5.25C5.22 4.147 6.68 3.5 8.006 3.5c1.397 0 2.673.73 2.673 2.24 0 1.08-.635 1.594-1.244 2.057-.737.559-1.01.768-1.01 1.486v.355H7.117l-.007-.463c-.038-.927.495-1.498 1.168-1.987.59-.444.965-.736.965-1.371 0-.825-.628-1.168-1.314-1.168-.901 0-1.358.603-1.358 1.384zm1.251 6.443c-.584 0-1.009-.394-1.009-.927 0-.552.425-.94 1.01-.94.609 0 1.028.388 1.028.94 0 .533-.42.927-1.029.927z"
                fillRule="evenodd"
              />
            </svg>
          </span>
          <div
            className="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-top type-dark"
            data-id="tooltip"
            id="UploadPhotos"
          >
            <style
              dangerouslySetInnerHTML={
                Object {
                  "__html": "
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  ",
                }
              }
            />
            Maximum 6 pictures
          </div>
        </span>
        <div
          className="flex justify-center px-6 pt-5 pb-6 border-2 border-gray-300 border-dashed rounded-md"
        >
          <div
            className="space-y-1 text-center"
          >
            <svg
              aria-hidden="true"
              className="w-12 h-12 mx-auto text-gray-400"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 48 48"
            >
              <path
                d="M28 8H12a4 4 0 00-4 4v20m32-12v8m0 0v8a4 4 0 01-4 4H12a4 4 0 01-4-4v-4m32-4l-3.172-3.172a4 4 0 00-5.656 0L28 28M8 32l9.172-9.172a4 4 0 015.656 0L28 28m0 0l4 4m4-24h8m-4-4v8m-12 4h.02"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
              />
            </svg>
            <input
              accept="image/png, image/jpg, image/jpeg, image/JPG, image/JPEG"
              className="photo-upload-width photo-upload hover:text-indigo-500"
              name="logo"
              onChange={[Function]}
              type="file"
            />
            <p
              className="text-xs text-gray-500"
            >
              PNG or JPG up to 2MB
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <label>
    <input
      className="btn btn-green form-btn w-1/2"
      onClick={[Function]}
      type="submit"
    />
  </label>
</div>
`;
